src/lit-element.ts

```diff
@@ -11,286 +11,34 @@
-   * returning an element into which to render. By default this creates
-   * a shadowRoot for the element. To render into the element's childNodes,
-   * return `this`.
-   * @returns {Element|DocumentFragment} Returns a node into which to render.
-   */
-  protected _createRoot(): Element|DocumentFragment {
-    return this.attachShadow({mode : 'open'});
-  }
-
-  /**
-   * Override which returns the value of `_shouldRender` which users
-   * should implement to control rendering. If this method returns false,
-   * _propertiesChanged will not be called and no rendering will occur even
-   * if property values change or `requestRender` is called.
-   * @param _props Current element properties
-   * @param _changedProps Changing element properties
-   * @param _prevProps Previous element properties
-   * @returns {boolean} Default implementation always returns true.
-   */
-  _shouldPropertiesChange(_props: object, _changedProps: object,
-                          _prevProps: object): boolean {
-    const shouldRender = this._shouldRender(_props, _changedProps, _prevProps);
-    if (!shouldRender && this.__resolveRenderComplete) {
-      this.__resolveRenderComplete(false);
-    }
-    return shouldRender;
-  }
+export abstract class LitElement extends UpdatingElement {
 
   /**
-   * Implement to control if rendering should occur when property values
-   * change or `requestRender` is called. By default, this method always
-   * returns true, but this can be customized as an optimization to avoid
-   * rendering work when changes occur which should not be rendered.
-   * @param _props Current element properties
-   * @param _changedProps Changing element properties
-   * @param _prevProps Previous element properties
-   * @returns {boolean} Default implementation always returns true.
+   * Override which performs element rendering by calling the `render` method.
+   * Override to perform tasks before and/or after updating.
    */
-  protected _shouldRender(_props: object, _changedProps: object,
-                          _prevProps: object): boolean {
-    return true;
-  }
-
-  /**
-   * Override which performs element rendering by calling
-   * `_render`, `_applyRender`, and finally `_didRender`.
-   * @param props Current element properties
-   * @param changedProps Changing element properties
-   * @param prevProps Previous element properties
-   */
-  _propertiesChanged(props: object, changedProps: object, prevProps: object) {
-    super._propertiesChanged(props, changedProps, prevProps);
-    const result = this._render(props);
-    if (result && this._root !== undefined) {
-      this._applyRender(result, this._root!);
-    }
-    this._didRender(props, changedProps, prevProps);
-    if (this.__resolveRenderComplete) {
-      this.__resolveRenderComplete(true);
+  protected update(_props: PropertyValues) {
+    super.update(_props);
+    if (typeof this.render === 'function') {
```

Comment:
Why not implement a default `render` method that throws an error? By doing that, we can avoid this `if` check. We can prevent users from calling `super.render` by documenting the default behavior.

Reply:
A method can't both be abstract and have an implementation. For TypeScript users, they'll get an error in their IDE that they should implement, which is better because it's sooner.